ql/src/java/org/apache/hadoop/hive/ql/Driver.java

```diff
@@ -515,14 +516,16 @@ private void preparForCompile(boolean resetTaskIds) throws CommandProcessorExcep
   }
 
   private void prepareContext() throws CommandProcessorException {
+    String originalCboInfo = context != null ? context.cboInfo : null;
```

Comment:
I feel that this save/load is a bit out-of-scope here;
I think that it might be possible to move this into the `ReCompilePlugin#Hook` - what do you think?

Reply:
The issue here is that the context we prepared in the `ReCompilePlugin#Hook` will be closed here, and a new one will be recreated:
```
    if (context != null && context.getExplainAnalyze() != AnalyzeState.RUNNING) {
      // close the existing ctx etc before compiling a new query, but does not destroy driver
      closeInProcess(false);
    }

    if (context == null) {
      context = new Context(driverContext.getConf());
    }
```

We have to save the cbo info pushed from the `ReCompilePlugin#Hook`

Reply:
Sadly this happens inside the `coreDriver.compileAndRespond(statement)` so even if we move the change to the `beforeCompile` the context will be recreated in the `prepareContext()`